go/vt/vtgate/engine/dbddl.go

```diff
@@ -0,0 +1,188 @@
+// The default behaviour is to just return an error
+type DBDDL struct {
+	name         string
+	create       bool
+	queryTimeout int
+
+	noInputs
+	noTxNeeded
+}
+
+// NewDBDDL creates the engine primitive
+// `create` will be true for CREATE, and false for DROP
+func NewDBDDL(dbName string, create bool, timeout int) *DBDDL {
+	return &DBDDL{
+		name:         dbName,
+		create:       create,
+		queryTimeout: timeout,
+	}
+}
+
+// RouteType implements the Primitive interface
+func (c *DBDDL) RouteType() string {
+	if c.create {
+		return "create database"
+	}
+	return "drop database"
+}
+
+// GetKeyspaceName implements the Primitive interface
+func (c *DBDDL) GetKeyspaceName() string {
+	return c.name
+}
+
+// GetTableName implements the Primitive interface
+func (c *DBDDL) GetTableName() string {
+	return ""
+}
+
+// Execute implements the Primitive interface
+func (c *DBDDL) Execute(vcursor VCursor, _ map[string]*querypb.BindVariable, _ bool) (*sqltypes.Result, error) {
+	name := vcursor.GetDBDDLPluginName()
+	plugin, ok := databaseCreatorPlugins[name]
+	if !ok {
+		log.Errorf("'%s' database ddl plugin is not registered. Falling back to default plugin", name)
+		plugin = databaseCreatorPlugins[defaultDBDDLPlugin]
+	}
+	if c.queryTimeout != 0 {
+		cancel := vcursor.SetContextTimeout(time.Duration(c.queryTimeout) * time.Millisecond)
+		defer cancel()
+	}
+
+	ctx := vcursor.Context()
+	if c.create {
+		err := plugin.CreateDatabase(ctx, c.name)
+		if err != nil {
+			return nil, err
+		}
+		var destinations []*srvtopo.ResolvedShard
+		for {
+			// loop until we have found a valid shard
+			destinations, _, err = vcursor.ResolveDestinations(c.name, nil, []key.Destination{key.DestinationAllShards{}})
+			if err == nil {
+				break
+			}
+			log.Errorf("waiting for topo to be updated after create db: %s", err.Error())
+			select {
+			case <-ctx.Done(): //context cancelled
+				return nil, vterrors.Errorf(vtrpc.Code_DEADLINE_EXCEEDED, "could not validate created database")
+			case <-time.After(500 * time.Millisecond): //timeout
+			}
+		}
+		var queries []*querypb.BoundQuery
+		for range destinations {
+			queries = append(queries, &querypb.BoundQuery{
+				Sql:           "select 42 from dual where null",
+				BindVariables: nil,
+			})
+		}
+
+		for {
+			_, errors := vcursor.ExecuteMultiShard(destinations, queries, false, true)
+
+			noErr := true
+			for _, err := range errors {
+				if err != nil {
+					noErr = false
+					log.Errorf("waiting for healthy tablets to be available after create db: %s", err.Error())
+					select {
+					case <-ctx.Done(): //context cancelled
+						return nil, vterrors.Errorf(vtrpc.Code_DEADLINE_EXCEEDED, "could not validate created database")
+					case <-time.After(500 * time.Millisecond): //timeout
+					}
+					break
+				}
+			}
+			if noErr {
+				break
+			}
+		}
+		return &sqltypes.Result{RowsAffected: 1}, nil
+	}
+
+	err := plugin.DropDatabase(ctx, c.name)
```

Comment:
We had discussed that after the plugin says it's done, we should wait until vtgate observes that the keyspace is gone. The idea is to mimic the semantic in MySQL, which is that it's safe to re-create a database immediately after `DROP DATABASE` returns. So we should probably check whatever signal we check when handling `CREATE DATABASE` here and wait for that signal to say the keyspace doesn't exist. WDYT?

Reply:
> We had discussed that after the plugin says it's done, we should wait until vtgate observes that the keyspace is gone. The idea is to mimic the semantic in MySQL, which is that it's safe to re-create a database immediately after `DROP DATABASE` returns. So we should probably check whatever signal we check when handling `CREATE DATABASE` here and wait for that signal to say the keyspace doesn't exist. WDYT?

Just want to call it out that the current implementation depends on the plugin to take care of changing topo server metadata to tell us that keyspace is created or deleted. 
I have added the necessary checks to ensure that. 

Also you can change the query timeout with query like `create database /*vt+ QUERY_TIMEOUT_MS=30000 */ <database_name>`

Reply:
> Just want to call it out that the current implementation depends on the plugin to take care of changing topo server metadata to tell us that keyspace is created or deleted.

That works for me. I just wanted vtgate to take care of the question, "Can this vtgate instance see the keyspace yet/still?" since the plugin can't determine that from topo alone.